src/app/server/DefaultTermsAndConditionsProvider.cpp

```diff
@@ -0,0 +1,219 @@
+    2; // Extra space for rollback compatibility
+} // namespace
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::Init(
+    chip::PersistentStorageDelegate * const inPersistentStorageDelegate,
+    const chip::Optional<chip::app::TermsAndConditions> & inmRequiredAcknowledgements)
+{
+    VerifyOrReturnError(nullptr != inPersistentStorageDelegate, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mPersistentStorageDelegate = inPersistentStorageDelegate;
+    mRequiredAcknowledgements  = inmRequiredAcknowledgements;
+
+    if (CHIP_NO_ERROR == LoadAcceptance(mLatchedAcceptance))
+    {
+        mTemporalAcceptance.SetValue(mLatchedAcceptance.Value());
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::CheckAcceptance(const Optional<TermsAndConditions> & inTermsAndConditions,
+                                                                         TermsAndConditionsState & outState) const
+{
+    // No validation checks required if no required terms and conditions
+    if (!mRequiredAcknowledgements.HasValue())
+    {
+        outState = TermsAndConditionsState::OK;
+        return CHIP_NO_ERROR;
+    }
+
+    // Validate if we have received any terms and conditions acceptance
+    if (!inTermsAndConditions.HasValue())
+    {
+        ChipLogError(AppServer, "No terms and conditions have been accepted");
+        outState = TermsAndConditionsState::TC_ACKNOWLEDGEMENTS_NOT_RECEIVED;
+        return CHIP_NO_ERROR;
+    }
+
+    // Validate the accepted version first...
+    if (mRequiredAcknowledgements.Value().version > inTermsAndConditions.Value().version)
+    {
+        ChipLogProgress(AppServer, "Minimum terms and conditions version, 0x%04x, has not been accepted",
+                        mRequiredAcknowledgements.Value().version);
+        outState = TermsAndConditionsState::TC_MIN_VERSION_NOT_MET;
+        return CHIP_NO_ERROR;
+    }
+
+    // Validate the accepted bits second...
+    if (mRequiredAcknowledgements.Value().value != (mRequiredAcknowledgements.Value().value & inTermsAndConditions.Value().value))
+    {
+        ChipLogProgress(AppServer, "Required terms and conditions, 0x%04x,have not been accepted",
+                        mRequiredAcknowledgements.Value().value);
+        outState = TermsAndConditionsState::REQUIRED_TC_NOT_ACCEPTED;
+        return CHIP_NO_ERROR;
+    }
+
+    // All validation check succeeded...
+    outState = TermsAndConditionsState::OK;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::CommitAcceptance()
+{
+    if (!mTemporalAcceptance.HasValue())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    mLatchedAcceptance.SetValue(mTemporalAcceptance.Value());
+    return StoreAcceptance(mLatchedAcceptance);
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::GetAcceptance(Optional<TermsAndConditions> & outTermsAndConditions) const
+{
+    outTermsAndConditions = mTemporalAcceptance;
```

Comment:
This should check if temporary present. If not, return latched. Otherwise it seems like there could be a situation:

- Fail-safe started, no conditions set during the fail-safe
- Fail-safe expires, Revert is called --> mTemporal is now empty
- New fail safe started (to retry the commissioning operations)
  - HERE: GetAcceptance doesn't return the previously fully latched/committed value.

Please add a unit test for the above case ^^^


I would recommend that the init doesn't set the temporal (temporary) version and that the GetAcceptance always does the check between temporary and latched.